data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,610 @@
+
+Putting the parentheses the other way does not work:
+```ocaml
+# let g : ((int -> int) -> int) = ( + );;
+Error: This expression has type int -> int -> int
+       but an expression was expected of type (int -> int) -> int
+       Type int is not compatible with type int -> int
+```
+
+In mathematical language, the type arrow operator _associates to the right_. Function types without parentheses should be treated as if they have parentheses placed to the right in the same way that the type of `f` was declared above. These types are the same:
+- `int -> int -> int`
+- `int -> (int -> int)`
+
+But only the first is displayed, so it is not the same as `(int -> int) -> int`.
+
+Most importantly, this means a *binary* function isn't really one that takes two parameters. In the case of addition, it is a function that takes an integer and returns a function of type `int -> int`. And it is indeed!
+```ocaml
+# ( + ) 2;;
+- : int -> int = <fun>
+```
+
+Passing a single integer to the addition returns a function of type `int -> int`. That function value is a closure. The value passed as the first parameter, here 2, is captured as if it had been in an earlier definition.
+
+### Passing Tuples
+
+Consider this function:
+```ocaml
+# let space_cat (s, t) = s ^ " " ^ t;;
+val space_cat : string * string -> string = <fun>
+```
+
+It takes a pair of strings and returns a string.
+```ocaml
+# space_cat ("hello", "world");;
+- : string = "hello world"
+```
+
+Asking whether `space_cat` is a binary function or not makes sense. Morally, it takes two parameters: the two strings that sandwich the space character. Wrapped as a pair, they are a single piece of data, so `space_cat` has a single parameter. But it is that pair's elements that are relevant, not the pair as a whole. So morally, `space_cat` is a binary function. Syntactically, the call `space_cat ("hello", "world")` looks pretty much like a high school maths function call or a spreadsheet application, both having two parameters: `"hello"` and `"world"`.
+
+### Currying and Uncurrying
+
+In the two previous sections, two kinds of “multiple parameter” functions have been presented.
+- Functions returning a function, such as `( + )`
+- Functions taking a pair or a tuple as a parameter, such as `space_cat`
+
+Interestingly, both kinds of functions provide a way to pass several parameters while being functions with a single parameter. From this perspective, it makes sense to say: “All functions have a single argument.”
+
+This goes even further. Any function of the same kind can be translated into an equivalent function of the second kind, and conversely. Using OCaml support for higher-order functions, it is possible to define those transformations as functions.
```

Comment:
Is the "first" and "second" kind referring to the bullet list in li.504-505? If so, let's number that list.